### ISSUE TITLE
🚀 Feature: ESLint rule to encourage Component Design Guidelines

### ISSUE THREAD

**Issue by luchillo17** [backstage/backstage#14964]
### 🔖 Feature description

Make use of ESLint rules to encourage the [Component Design Guidelines](https://backstage.io/docs/dls/component-design-guidelines), specifically forbidding HTML tags that instead should use components from either `@backstage/core-components` or MUI components.

### 🎤 Context

Now that the guidelines have been defined, agreed and put publicly in the docs, we should stride towards applying it to the whole project, we could either add the rule and fix the whole project in one go, or request every package & plugin author to add the rule and fix their own workspace.

### ✌️ Possible Implementation

[eslint-plugin-react](https://github.com/jsx-eslint/eslint-plugin-react#list-of-supported-rules) has many rules we could use, but in this case I recommend the [forbid-elements](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/forbid-elements.md) rule.

Basically we blacklist common HTML tags that should be replaced, like `<div>` for MUI's `<Box>`, `<p>` and `<span>` for `<Typography>` and  `<Typography component="span">`, `<button>` for `<Button>`...

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by freben** [comment]
Hi!

I'm not entirely sure that the guidelines are so strict, that linting it possible for some of these. One that comes to mind is `div` vs `Box`. The wording should perhaps be clarified, because it's not clear to me whether they mean "never use raw div", or rather "remember to always use theme.spacing instead of raw values, and Box is an easier way than div to achieve this if you want".

Maybe _some_ things can be linted against. Not sure we'd legitimately need to use `<button>` for example, but there are a small number of those in the repo, mostly in tests.

**Comment by luchillo17** [comment]
It was indeed "never use raw div", the same for plain text, should always be wrapped in `Typography`, though not sure the proposed rule can check for that...

**Comment by luchillo17** [comment]
The button one was an example I saw when reading the rule docs I believe

**Comment by luchillo17** [comment]
Btw before finding that rule, I was about to make a custom ESLint rule, starting to learn how to explore the AST and define rules, hopefully this existing rule is enough...

**Comment by freben** [comment]
Alright, to move this forward a bit, would you like to try to start small? Perhaps adding a lint rule to use Button instead of button, to see how it goes. And it would not be in the backstage-cli's rules that it ships to everybody - that's too forceful to push upon everybody. Instead it would be in our own repo's root lint config, most likely, if it's possible (it mustn't overwrite the backstage-cli rules that we ALSO want to retain support for :))

**Comment by luchillo17** [comment]
Looked around one to ensure text was wrapped at some level by `Typography` but couldn't find a rule like that, and making one I feel like it wouldn't be very performant...

Anyway, want me to create a PR with that rule to see how many errors pop out? or want me to fix them in one go as well?

**Comment by luchillo17** [comment]
Here it is: https://github.com/backstage/backstage/pull/15037

**Comment by luchillo17** [comment]
Hmm there's more linting errors than I thought, in the PR it stops on the `app` package immediately, but in local run it spews a lot of linting errors, was not expecting `plugin:react/recommended` to throw so many errors, seems we're far from React's good practices (like using `children` prop instead of putting the children nested inside the component...).

Anyway, going to remove that and keep only the button rule.

**Comment by luchillo17** [comment]
Hmm there's less matches than expected, in a `yarn lint:all` only got the `button` lint error in these packages:
- `packages/core-components`: 3 warnings
- `plugins/techdocs`: 2 warnings
- `plugins/catalog-import`: 3 warnings

**Comment by luchillo17** [comment]
Which one would be a next candidate? paragraphs seem like a good one, I count around 20 instances with global search.

**Comment by jhaals** [comment]
paragraphs sounds good! Looks like it's happening in #15251 🎉

**Comment by luchillo17** [comment]
All PRs were merged successfully 🎉 🎉
